components/selectors/parser.rs

```diff
@@ -376,15 +378,40 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
 
 impl<Impl: SelectorImpl> ToCss for ComplexSelector<Impl> {
     fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
-        if let Some((ref next, ref combinator)) = self.next {
-            next.to_css(dest)?;
+       use smallvec::SmallVec;
+       let mut current = self;
+       let mut nodes = SmallVec::<[Arc<ComplexSelector<Impl>>; 8]>::new();
+
+       loop {
+           match current.next {
+               None => break,
+               Some((ref next, _)) => {
+                   current = &**next;
+                   nodes.push((*next).clone());
+               }
+           }
+       }
+
+       while let Some(v) = nodes.pop() {
+           if let Some((_, ref combinator)) = v.next {
+               combinator.to_css(dest)?;
+           }
+
+           for simple in &v.compound_selector {
+               simple.to_css(dest)?;
+           }
+       }
+
+        if let Some((_, ref combinator)) = self.next {
```

Comment:
There's no need to duplicate the logic for `self` here, let's just push ourselves to the list in the first place.